prepare_dataset.py

```diff
@@ -186,6 +193,48 @@ def resize_and_crop(data_prep_local_dir, target_size, image_cropping_params):
                                 (Path(data_prep_local_dir, 'resized', scan, 'annotations', scan_annotation_files[
                                     image_ind].name).as_posix()).replace('.', ('_crop' + str(counter_crop) + '.')))
                             counter_crop += 1
+                elif image_cropping_params['type'] == 'class':  # 'smart' crop: do not train with pad, some overlap okay (still aug'd)
+                    assert 'num_pos_per_class' in image_cropping_params
+                    assert 'num_neg_per_class' in image_cropping_params
+                    assert image_cropping_params['num_pos_per_class'] > 0  # logical choice
+                    assert image_cropping_params['num_pos_per_class'] <= 36  # suits 4600 x 2048 img with 512 x 512 target
+                    assert image_cropping_params['num_neg_per_class'] >= 0  # logical choice if 0
+                    assert image_cropping_params['num_neg_per_class'] <= 36  # suits 4600 x 2048 img with 512 x 512 target
+                    class_pos_min_threshold = 1  # min thresh that defines class pos pixel qty per crop
+                    for c, gvs_in_c in class_annotation_mapping.items():
+                        assert "class_" in c
+                        assert "_annotation_GVs" in c, "'_annotation_GVs' must be in the class name to indicate these are grayvalues"
+                        class_name = c[:-len('_annotation_GVs')]
+                        if np.size(np.asarray(annotation)[np.isin(np.asarray(annotation), gvs_in_c)]) >= class_pos_min_threshold:  # if class-pos is present in full annot
+                            for counter_classpos_crop in range(image_cropping_params['num_pos_per_class']):
+                                flag_crop_pass = 0
+                                counter_classpos_tries = 0  # getting this far presents no guarantees of ok crop selection
+                                while flag_crop_pass == 0 and counter_classpos_tries < 1000:  # stopgap soln:
```

Comment:
Same comment as above with the 1000

Reply:
Not sure if this is global enough, i moved it to the top of the function def
```def resize_and_crop(data_prep_local_dir, target_size, image_cropping_params, class_annotation_mapping):
    Path(data_prep_local_dir, 'resized').mkdir(parents=True, exist_ok=True)
    assert 'type' in image_cropping_params
    assert target_size[0] > 0
    assert target_size[1] > 0
    max_tries_rand_crop_per_class = 1000
    for scan in [p.name for p in Path(data_prep_local_dir, 'downsampled').iterdir()]:```

Reply:
question is, should this also be captured in metadata and therefore an input parameter in the dataset config file? 1000 is just an arbitrary number, tho maybe it's not too important